### PR TITLE
util/threadpool: add thread pool scheduler

### PR DESCRIPTION
**Pull request by overvenus**
Signed-off-by: Neil Shen <overvenus@gmail.com>

## What have you changed? (mandatory)

Add scheduler for thread pools, so we can spawn a task to thread pools from any threads we want.

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

Unit tests.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

**Comment by breeswish** [comment]
We have too many kind of schedulers. Let's rename it. Usually we call it sender, for example, in rust-threadpool and tokio-threadpool.

**Comment by overvenus** [comment]
> We have too many kind of schedulers.

Indeed, because we usually call it scheduler in TiKV, eg, `worker::Scheduler` and `worker::futures::Scheduler`.
I think it's fine as long as there is only one Scheduler for a module, like Sender in futures, `futures::sync::mpsc::Sender
futures::sync::oneshot::Sender
futures::unsync::mpsc::Sender
futures::unsync::oneshot::Sender`.
What do you think? @breeswish

**Comment by breeswish** [comment]
@overvenus Ok.

File: src/util/threadpool.rs [hunk 1]
```diff
@@ -150,6 +150,31 @@ struct ScheduleState<Ctx> {
...
+    task_count: Arc<AtomicUsize>,
+}
+
+impl<Ctx> Scheduler<Ctx> {
+    pub fn schedule<F>(&self, job: F)
+    where
+        F: FnOnce(&mut Ctx) + Send + 'static,
+        Ctx: Context,
+    {
+        let task = Task::new(job);
```
Comment (siddontang) [inline src/util/threadpool.rs]: the code is duplicated with https://github.com/tikv/tikv/blob/a8c1ccb1935ab20aa20168e2b57e71304f6c50ef/src/util/threadpool.rs#L238

**Pull request merged**